csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/NugetPackageRestorer.cs

```diff
@@ -674,10 +700,40 @@ private bool IsFeedReachable(string feed, int timeoutMilliSeconds, int tryCount,
             return (timeoutMilliSeconds, tryCount);
         }
 
+        /// <summary>
+        /// Checks that we can connect to all Nuget feeds that are explicitly configured in configuration files.
+        /// </summary>
+        /// <param name="explicitFeeds">Outputs the set of explicit feeds.</param>
+        /// <returns>True if all feeds are reachable or false otherwise.</returns>
         private bool CheckFeeds(out HashSet<string> explicitFeeds)
         {
-            logger.LogInfo("Checking Nuget feeds...");
             (explicitFeeds, var allFeeds) = GetAllFeeds();
+            HashSet<string> feedsToCheck = explicitFeeds;
+
+            // If private package registries are configured for C#, then check those
+            // in addition to the ones that are configured in `nuget.config` files.
+            this.dependabotProxy?.RegistryURLs.ForEach(url => feedsToCheck.Add(url));
+
+            var allFeedsReachable = this.CheckFeeds(feedsToCheck);
```

Comment:
If the dependabot proxy is set, we also explicitly add the public nuget org feed as a source - so we should probably check whether that feed is reachable as well.

Reply:
I have changed this so that the public feed is not manually added when private registries are configured in https://github.com/github/codeql/pull/18850/commits/4d3b0246b5c2da3e86129cbb1232baac0cb68a6f, since `explicitFeeds` should already contain it. I checked and `dotnet nuget list source` does return it unless there is a `<clear />` entry in the list of `packageSources` in `nuget.config`.

Reply:
Actually, there's more subtly to this. `GetNugetFeeds` doesn't include it, because `dotnet nuget list source` only includes the feeds that are explicitly configured in a given configuration file if it is given an `--configfile` argument. `GetNugetFeedsFromFolder` does include the public feed (unless there's a `<clear />` element), but we don't check whether these "inherited" feeds are reachable already. 

So, we didn't previously check that the public feed is reachable anyway. @tamasvajk left a comment saying that we could check that they are reachable, but don't because of authentication requirements(?). 

I think the real lesson for my changes here is that I should really use `allFeeds` as argument for `RestoreProjects` rather than just `explicitFeeds`.

Reply:
OK that also still doesn't quite work with the current logic, because `GetAllFeeds` only invokes `GetNugetFeedsFromFolder` in folders that contain `nuget.config` files to begin with. In our case, we may not have a `nuget.config` file at all, so `GetNugetFeedsFromFolder` never gets called, even though it would give us the public feed as expected.

Reply:
If the reachability of the public nuget feed wasn't checked before, then maybe there is no need to check it now either.
However, I think its availability is checked when it is added as a fallback feed (not as an inherited feed).

Reply:
> If the reachability of the public nuget feed wasn't checked before, then maybe there is no need to check it now either.

In any case, it's probably a separate discussion to the changes in this PR. I have also now removed that part of the code that manually added the public feed in https://github.com/github/codeql/pull/18850/commits/be95d335b7286a91ba743390cf5c8480d788ebeb.

> However, I think its availability is checked when it is added as a fallback feed (not as an inherited feed).

That's true.